documentation/projects/proposals/popularity_optimizations/20230420-implementation_plan_popularity_optimizations.md

```diff
@@ -0,0 +1,486 @@
+views, etc). Because of this, the constant needs to be periodically updated for
+accuracy as more data is consumed.
+
+Refreshing this view results in the constants being recalculated. We currently
+do this once a month. It must also be done whenever a new metric is added.
+
+#### `image_view`
+
+A materialized view sourced from the `image` table, which adds the calculated
+`standardized_popularity` column. This normalized popularity score is calculated
+by a simple operation using the raw score and the popularity constant.
+
+The effect of refreshing the view is to:
+
+- Add all images which were ingested since the last time the view was refreshed,
+  and calculate their popularity
+- Update existing records which may have been modified on the `image` table
+  since the last refresh
+- Update the `standardized_popularity` for all records to use _the current
+  popularity constant_
+
+In a data refresh, the data from this view is copied into a new table in the API
+DB, which is used to build new Elasticsearch indices and eventually swapped with
+the live tables. As a result, the view **must** be refreshed before the data
+refresh can run, or newly ingested data would not make it to the API.
+
+## Outlined Steps
+
+<!-- Describe the implementation step necessary for completion. -->
+
+The basic strategy proposed here for decoupling popularity calculations from the
+data refresh is:
+
+1. Add `standardized_popularity` as a `nullable` column on the `image` and
+   `audio` tables. ([link](#add-standardized-popularity-to-the-media-tables))
+1. Update the `provider_dag_factory` to calculate `standardized_popularity` **at
+   ingestion**.
+   ([link](#Update-provider-DAGs-to-calculate-standardized-popularity))
+1. Create a new `popularity_refresh` DAG
+   ([link](#create-a-popularity_refresh-dag)) which:
+   1. Recalculates the popularity constants
+   1. Updates the `standardized_popularity` score for existing records, using
+      the new constants.
+1. Remove the popularity calculation steps from the data refresh DAGs.
+   ([link](#remove-popularity-steps-from-the-data_refresh-dags))
+1. Begin running the data refresh directly from the media tables, and drop the
+   materialized views. ([link](#run-the-data-refresh-from-the-media-tables))
+
+More details about each of these steps, and the order in which they must be
+implemented, is included in the following sections.
+
+### Add standardized popularity to the media tables
+
+This step is simple:
+
+```pgsql
+ALTER TABLE image ADD COLUMN standardized_popularity DOUBLE PRECISION;
+ALTER TABLE audio ADD COLUMN standardized_popularity DOUBLE PRECISION;
+```
+
+The column can (and indeed _must_) be nullable, as not all providers support
+popularity data. Because this configuration exactly matches the schema of the
+current `image_view` and `audio_view` matviews, we will be able to begin running
+data refreshes directly off these tables with very few changes in the ingestion
+server.
```

Comment:
Is it possible to do that without running into the concurrency/deadlocking problems with simultaneous updates coming from the provider DAGs?

I suppose one thing we could do is do the update one provider at a time, and temporarily pause the provider DAG while the update runs (for this initial run).

Reply:
We should definitely wait at least until this IP is fully approved before starting something like that, just to make sure we're going with this approach -- but that's a really interesting idea if we can make it work. It would allow us to enable the data refreshes more quickly, _and_ let us do a full production trial of the refresh-from-media-tables before investing the time in the rest of the work 😮

Reply:
Oh good call, agreed on all fronts!